File: qpid/java/perftests/src/main/java/org/apache/qpid/client/latency/MessageConsumer.java
/* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.qpid.client.latency;

import org.apache.qpid.client.AMQConnection;
import org.apache.qpid.client.AMQQueue;
import org.apache.qpid.client.AMQTopic;
import org.apache.qpid.client.perf.Options;
import org.apache.qpid.framing.AMQShortString;
import org.apache.qpid.requestreply.InitialContextHelper;

import javax.jms.*;
import javax.naming.Context;

/**
 *
 *
 */
public class MessageConsumer  extends Options  implements MessageListener 
{
    private javax.jms.MessageProducer _producer;
    private AMQConnection _connection;
    private final Object _lock = new Object();
    private Session _session;
    private int _receivedMessages = 0;
    private long _timeFirstMessage;
    private long _timeLastMessage;
   private void init()
    {
        this.parseOptions();
        try
        {
            Context context = InitialContextHelper.getInitialContext("");
            ConnectionFactory factory = (ConnectionFactory) context.lookup("local");
             _connection = (AMQConnection) factory.createConnection("guest","guest");
             _session = _connection.createSession(_transacted, Session.AUTO_ACKNOWLEDGE);
              Destination dest = Boolean.getBoolean("useQueue")? (Destination) context.lookup("testQueue") :
            (Destination) context.lookup("testTopic");
            Destination syncQueue   = (Destination) context.lookup("syncQueue");
            _producer = _session.createProducer(syncQueue);
            // this should speedup the message producer
            _producer.setDisableMessageTimestamp(true);
            javax.jms.MessageConsumer consumer = _session.createConsumer(dest);
            consumer.setMessageListener(this);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    private void run()
    {
        try
        {
           synchronized(_lock)
           {
               _connection.start();
               try
               {
                   _lock.wait();
               }
               catch (InterruptedException e)
               {
                   e.printStackTrace();
               }
           }
            // send sync message;
           _producer.send(_session.createMessage());
            System.out.println("Time to receive " + _logFrequency + " messages is: " + (_timeLastMessage - _timeFirstMessage) );
            double rate =  _logFrequency /  ((_timeLastMessage - _timeFirstMessage)  *1.0) *1000 ;
             System.out.println("The rate is " + rate  + " msg/s" );
            double latency =  ((_timeLastMessage - _timeFirstMessage)  *1.0) / _logFrequency;
            System.out.println("The latency is " + latency  + " milli secs" );
            _connection.close();
        }
        catch (JMSException e)
        {
            e.printStackTrace();
        }
    }

   public void onMessage(Message message)
    {
        if( _receivedMessages == 0)
        {
            _timeFirstMessage = System.currentTimeMillis();
        }
        _receivedMessages++;      
        if( _receivedMessages == _logFrequency)
        {
            _timeLastMessage = System.currentTimeMillis();
            synchronized(_lock)
            {
                _lock.notify();
            }
        }
    }

    public static void main(String[] args)
    {
        try
        {
            MessageConsumer test = new MessageConsumer();
            test.init();
            test.run();
        }
        catch(Exception e)
        {
            e.printStackTrace();
        }
    }

}

File: cpp/src/tests/storePerftools/asyncPerf/MessageConsumer.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * \file MessageConsumer.cpp
 */

#include "MessageConsumer.h"

#include "TestOptions.h"

#include "qpid/broker/SimpleDeliveryRecord.h"
#include "qpid/broker/SimpleQueue.h"
#include "qpid/broker/SimpleTxnAccept.h"
#include "qpid/broker/SimpleTxnBuffer.h"

#include <stdint.h> // uint32_t

namespace tests {
namespace storePerftools {
namespace asyncPerf {

MessageConsumer::MessageConsumer(const TestOptions& perfTestParams,
                                 qpid::broker::AsyncStore* store,
                                 qpid::broker::AsyncResultQueue& arq,
                                 boost::shared_ptr<qpid::broker::SimpleQueue> queue) :
        m_perfTestParams(perfTestParams),
        m_store(store),
        m_resultQueue(arq),
        m_queue(queue),
        m_stopFlag(false)
{}

MessageConsumer::~MessageConsumer() {}

void
MessageConsumer::record(boost::shared_ptr<qpid::broker::SimpleDeliveryRecord> dr) {
    m_unacked.push_back(dr);
}

void
MessageConsumer::commitComplete() {}

void
MessageConsumer::stop() {
    m_stopFlag = true;
}

void*
MessageConsumer::runConsumers() {
    const bool useTxns = m_perfTestParams.m_deqTxnBlockSize > 0U;
    uint16_t opsInTxnCnt = 0U;
    qpid::broker::SimpleTxnBuffer* tb = 0;
    if (useTxns) {
        tb = new qpid::broker::SimpleTxnBuffer(m_resultQueue);
    }

    uint32_t msgsPerConsumer = m_perfTestParams.m_numEnqThreadsPerQueue * m_perfTestParams.m_numMsgs /
                               m_perfTestParams.m_numDeqThreadsPerQueue;
    uint32_t numMsgs = 0UL;
    while (numMsgs < msgsPerConsumer && !m_stopFlag) {
        if (m_queue->dispatch(*this)) {
            ++numMsgs;
            if (useTxns) {
                // --- Transactional dequeue ---
                boost::shared_ptr<qpid::broker::SimpleTxnAccept> ta(new qpid::broker::SimpleTxnAccept(m_unacked));
                m_unacked.clear();
                tb->enlist(ta);
                if (++opsInTxnCnt >= m_perfTestParams.m_deqTxnBlockSize) {
                    tb->commitLocal(m_store);
                    if (numMsgs < msgsPerConsumer) {
                        tb = new qpid::broker::SimpleTxnBuffer(m_resultQueue);
                    }
                    opsInTxnCnt = 0U;
                }
            } else {
                // --- Non-transactional dequeue ---
                for (std::deque<boost::shared_ptr<qpid::broker::SimpleDeliveryRecord> >::iterator i = m_unacked.begin(); i != m_unacked.end(); ++i) {
                    (*i)->accept();
                }
                m_unacked.clear();
            }
        } else {
            ::usleep(1000); // TODO - replace this poller with condition variable
        }
    }

    if (opsInTxnCnt && !m_stopFlag) {
        tb->commitLocal(m_store);
    }

    return reinterpret_cast<void*>(0);
}

//static
void*
MessageConsumer::startConsumers(void* ptr) {
    return reinterpret_cast<MessageConsumer*>(ptr)->runConsumers();
}

}}} // namespace tests::storePerftools::asyncPerf
